text/0000-build-std.md

```diff
@@ -0,0 +1,4582 @@
+and build scripts ([?][rationale-sysroot-for-host-deps]). Artifact dependencies
+use the same standard library as the rest of the crate (pre-built or
+newly-built, as appropriate).
+
+*See the following sections for rationale/alternatives:*
+
+- [*Why put `build-std` in the Cargo config?*][rationale-build-std-in-config]
+- [*Why not always rebuild when the profile changes?*][rationale-why-not-always-rebuild]
+- [*Why have different build-std defaults depending on the profile?*][rationale-different-defaults]
+- [*Why rebuild the standard library automatically?*][rationale-why-automatic]
+- [*Why use the lockfile of the `rust-src` component?*][rationale-lockfile]
+- [*Why not build the standard library in incremental?*][rationale-incremental]
+- [*Why not produce a `dylib` for the standard library?*][rationale-no-dylib]
+- [*Why use the pre-built standard library for procedural macros and build-scripts?*][rationale-sysroot-for-host-deps]
+
+*See the following sections for relevant unresolved questions:*
+
+- [*Where should the `build-std` configuration in `.cargo/config` be and what should it be called?*][unresolved-config-location-name]
+- [*What should the values of the `build-std` config be named?*][unresolved-config-values]
+
+### Profiles
+[profiles]: #profiles
+
+Cargo will assume that the pre-built standard library matches the standard
+library's release profile ([?][rationale-assume-release-profile]). If the user
+changes the default release profile or builds with a different profile then this
+could trigger a rebuild of the standard library
+([?][rationale-ship-debug-std]), depending on the value of the `build-std`
+config as above.
+
+User's Cargo profile | Target modifiers changed? | Standard library profile
+-------------------- | ------------------------- | ------------------------
+`release`            | No                        | `release` (pre-built)
+`dev`                | No                        | `release` (pre-built)
+`release`            | Yes                       | `release` (newly built)
+`dev`                | Yes                       | `dev` (newly built)
+
+> [!NOTE]
+>
+> If `build-std` is set to `target-modifiers`, Cargo must decide if `build-std`
+> should be enabled. rustc will add a `--print target-modifiers` flag which will
+> print all of the flags treated as target modifiers, like `-Zretpoline`, with
+> one flag per line and its default value.
+>
+> If changing a profile configuration would result in one of these flags being
+> emitted by Cargo then it assumes a target modifier has changed from the
+> default release profile and would no longer match the pre-built standard
+> library.
+
+Unlike other dependencies, the profiles defined in the standard library's
+workspace will apply to its build even when used as a dependency
+([?][rationale-respect-std-profile]) and the user's profile configuration will
+override it only where the user's profile differs from its default.
```

Comment:
I'm worried about all these mechanisms tied to profiles is adding a large amount of special-case code into cargo. I don't have any actionable advice here, just expressing a worry that this is going to be difficult.